### PR TITLE
Add batch functionality for migrating data into ODPS

### PR DESCRIPTION
**Pull request by onesuper**
In the modified code, the **PrepareStatement** restores the newly added parameter into `batchedRows ` and pushes to ODPS via tunnel API in batch. All the type information are reserved in  `batchedRows` to represent the tunnel records.

Currently, SQLs like `INSERT INTO table (field1, field2) VALUES (?, ?);` will be denied by the implementation (will support later).

I add an example named `DataDumpBenchmark' to show how to use this functionality. It downloads from ODPS a 3.2GB-sized table (m_instance) while uploading to ODPS and costs 17 minutes in total.

### CODE REVIEW

File: src/main/java/com/aliyun/odps/jdbc/OdpsPreparedStatement.java [hunk 1]
```diff
@@ -81,6 +133,170 @@ public void clearParameters() throws SQLException {
   }
 
   @Override
+  public void clearBatch() throws SQLException {
+    batchedRows.clear();
```
Comment (lyman) [inline src/main/java/com/aliyun/odps/jdbc/OdpsPreparedStatement.java]: 这里需不需要把 parameters 也 clear 一下？

File: src/main/java/com/aliyun/odps/jdbc/OdpsPreparedStatement.java [hunk 2]
```diff
@@ -42,27 +45,71 @@
...
    */
-  private final HashMap<Integer, String> parameters = new HashMap<Integer, String>();
+  private HashMap<Integer, Object> parameters = new HashMap<Integer, Object>();
+
+  // When addBatch(), compress the parameters into a row
+  private List<Object[]> batchedRows = new ArrayList<Object[]>();
 
   OdpsPreparedStatement(OdpsConnection conn, String sql) {
     super(conn);
     this.sql = sql;
```
Comment (lyman) [inline src/main/java/com/aliyun/odps/jdbc/OdpsPreparedStatement.java]: Seems more reasonable to do verify logic here. I'm not sure if throw exception in constructor is a good idea. However, if not verified, it is meaningless for user to do any further action, such as setXXX and addBatch.

File: src/main/java/com/aliyun/odps/jdbc/OdpsPreparedStatement.java [hunk 3]
```diff
@@ -81,6 +133,170 @@ public void clearParameters() throws SQLException {
...
+      if (valNum != colNum) {
+        throw new SQLException(
+            "the table has " + colNum + " columns, but insert " + valNum + " values");
+      }
+
+      long startTime = System.currentTimeMillis();
+      TunnelRecordWriter recordWriter = (TunnelRecordWriter) session.openRecordWriter(0, true);
+      for (int i = 0; i < batchedSize; i++) {
+        Object[] row = batchedRows.get(i);
+        for (int col = 0; col < colNum; col++) {
```
Comment (lyman) [inline src/main/java/com/aliyun/odps/jdbc/OdpsPreparedStatement.java]: Can we just do an `ArrayRecord.set(Object[])` here?

File: src/main/java/com/aliyun/odps/jdbc/OdpsPreparedStatement.java (hunk 2 above)
Comment (onesuper) [inline src/main/java/com/aliyun/odps/jdbc/OdpsPreparedStatement.java]: 一种是干脆禁止除了 Insert into 以外的所有语句。只用 PS 批量上传。可以 OdpsPrepareStatement 让提供一个 acceptStatement() 方法，在 Connection.prepareStatement 里先 check 一下，如果 false 就抛异常。

一旦禁止其他 sql 以后，我们又会把 insert 语句都替换了 tunnel 上传，Prepare.executeQuery() 也都要禁掉。但有以下场景用户可能又需要用到：

```
String sql = "select * from people where firstname=? and lastname=?";
```

File: src/main/java/com/aliyun/odps/jdbc/OdpsPreparedStatement.java (hunk 3 above)
Comment (onesuper) [inline src/main/java/com/aliyun/odps/jdbc/OdpsPreparedStatement.java]: yep, trade correctness for performance. Currently, CPU time is not the bottleneck.

Comment (onesuper) [inline src/main/java/com/aliyun/odps/jdbc/OdpsPreparedStatement.java]: not sure whether ArrayRecord.set() checks the type correctness?

**Pull request merged**